homeassistant/components/mqtt/discovery.py

```diff
@@ -239,59 +365,125 @@ def async_discovery_message_received(msg: ReceiveMessage) -> None:  # noqa: C901
 
         component, node_id, object_id = match.groups()
 
-        if payload:
+        discovered_components: list[MqttComponentConfig] = []
+        if component == CONF_DEVICE:
+            # Process device based discovery message and regenerate
+            # cleanup config for the all the components that are being removed.
+            # This is done when a component in the device config is omitted and detected
+            # as being removed, or when the device config update payload is empty.
+            # In that case this will regenerate a cleanup message for all every already
+            # discovered components that were linked to the initial device discovery.
+            device_discovery_payload = _parse_device_payload(
+                hass, payload, object_id, node_id
+            )
+            if not device_discovery_payload:
+                return
+            device_config: dict[str, Any]
+            origin_config: dict[str, Any] | None
+            component_configs: dict[str, dict[str, Any]]
+            device_config = device_discovery_payload[CONF_DEVICE]
+            origin_config = device_discovery_payload.get(CONF_ORIGIN)
+            component_configs = device_discovery_payload[CONF_COMPONENTS]
+            for component_id, config in component_configs.items():
+                component = config.pop(CONF_PLATFORM)
+                # The object_id in the device discovery topic is the unique identifier.
+                # It is used as node_id for the components it contains.
+                component_node_id = object_id
+                # The component_id in the discovery playload is used as object_id
+                # If we have an additional node_id in the discovery topic,
+                # we extend the component_id with it.
+                component_object_id = (
+                    f"{node_id} {component_id}" if node_id else component_id
+                )
+                _replace_all_abbreviations(config)
```

Comment:
This is already called from `_parse_device_payload`, why do we do it again here for each component?
Can't we add a loop in `_replace_all_abbreviations` instead?

Reply:
I have moved this to _parse_device_payload